### PR TITLE
[pipeline-in-pod] Support parallel tasks

### PR DESCRIPTION
**Pull request by lbernick**
# Changes
This commit adds support for parallel tasks in the same pod. It writes step output to
/tekton/run/<pipeline_task_name>/<step_number> and supports multiple waitFiles per container.
It also pins the image versions for the entrypoint binary and git-init, to avoid breaking the controller.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [x] Commit messages includes a project tag in the subject - e.g. [OCI], [hub], [results], [task-loops]

_See [the contribution guide](https://github.com/tektoncd/experimental/blob/master/CONTRIBUTING.md)
for more details._

### CODE REVIEW

File: pipeline-in-pod/pkg/reconciler/pipelineinpod/entrypoint.go [hunk 1]
```diff
@@ -165,14 +183,54 @@ func orderContainers(commonExtraEntrypointArgs []string, ptcs []pipelineTaskCont
...
+}
+
+func getVolumesForStep(ptc pipelineTaskContainers, i int, runAfter []pipelineTaskContainers) ([]corev1.Volume, []corev1.VolumeMount) {
+	var volumes []corev1.Volume
+	s := ptc.containers[i]
+
+	// TODO (maybe) creds-init
+
+	// Add /tekton/run state volumes.
+	// Each step should only mount their own volume as RW,
```
Comment (dibyom) [inline pipeline-in-pod/pkg/reconciler/pipelineinpod/entrypoint.go:205]: Nice!

**Comment by dibyom** [comment]
Nice!
/approve

Just making sure I understand the volumeMount logic:
1. Each step mounts each volume for that pipeline task - volumes for other steps are mounted as RO while volumes for that particular step gets mounted as RW
2. Volumes from other pipeline tasks are not mounted at all so should not be visible to the step

Question: Would it be possible to share the same `emptyDir` or similar non persistent volume across pipelineTasks somehow with this implementation?

**Comment by lbernick** [comment]
> Just making sure I understand the volumeMount logic:
> 
> 1. Each step mounts each volume for that pipeline task - volumes for other steps are mounted as RO while volumes for that particular step gets mounted as RW

Yes

> 2. Volumes from other pipeline tasks are not mounted at all so should not be visible to the step

Not exactly; the first step in each task also mounts volumes for the last steps of any other tasks it has a `runAfter` dependency on, so it can know when those steps have completed.

Unclear what the security implications are here-- we'd probably have to prevent non tekton code from writing to the /tekton/run volumes somehow? but to some extent the previous tasks need to communicate in some way that they have finished

> Question: Would it be possible to share the same `emptyDir` or similar non persistent volume across pipelineTasks somehow with this implementation?

I'm not sure exactly what you mean by this, are you talking about a workspace shared among pipeline tasks or are you talking about using one volume for all of the /tekton/run data or other tekton-internal volumes that are mounted?

**Comment by lbernick** [comment]
> Just making sure I understand the volumeMount logic:
> 
> 1. Each step mounts each volume for that pipeline task - volumes for other steps are mounted as RO while volumes for that particular step gets mounted as RW

Yes

> 2. Volumes from other pipeline tasks are not mounted at all so should not be visible to the step

Not exactly; the first step in each task also mounts volumes for the last steps of any other tasks it has a `runAfter` dependency on, so it can know when those steps have completed.

Unclear what the security implications are here-- we'd probably have to prevent non tekton code from writing to the /tekton/run volumes somehow? but to some extent the previous tasks need to communicate in some way that they have finished

> Question: Would it be possible to share the same `emptyDir` or similar non persistent volume across pipelineTasks somehow with this implementation?

I'm not sure exactly what you mean by this, are you talking about a workspace shared among pipeline tasks or are you talking about using one volume for all of the /tekton/run data or other tekton-internal volumes that are mounted?

**Comment by dibyom** [comment]
> Not exactly; the first step in each task also mounts volumes for the last steps of any other tasks it has a runAfter dependency on, so it can know when those steps have completed.

Got it, that makes sense. I agree re: the security implications. 

> I'm not sure exactly what you mean by this, are you talking about a workspace shared among pipeline tasks or are you talking about using one volume for all of the /tekton/run data or other tekton-internal volumes that are mounted?

The former - I'd like a workspace that I can use throughout the pipeline.

**Comment by lbernick** [comment]
> > I'm not sure exactly what you mean by this, are you talking about a workspace shared among pipeline tasks or are you talking about using one volume for all of the /tekton/run data or other tekton-internal volumes that are mounted?
> 
> The former - I'd like a workspace that I can use throughout the pipeline.

This PR should only affect the /tekton/run volume mounts, not workspace volumes. This custom task should already support a workspace available for the whole pipeline.

**Comment by lbernick** [comment]
@jerop @dibyom @vdemeester @afrittoli if any of you are able to LGTM that would unblock me for some work I'm trying to get done on this project! 🙏

**Review by vdemeester — APPROVED** [review]
/lgtm

**Pull request merged**